appinventor/components/src/com/google/appinventor/components/runtime/BarChartDataModel.java

```diff
@@ -133,13 +134,40 @@ public void addTimeEntry(YailList tuple) {
 
     // If the entry count of the Data Series entries exceeds
     // the maximum allowed time entries, then remove the first one
-    if (getDataset().getEntryCount() >= maximumTimeEntries) {
-      getDataset().getValues().remove(0);
+    if (entries.size() >= maximumTimeEntries) {
+      entries.remove(0);
     }
 
-    getDataset().addEntry((BarEntry)getEntryFromTuple(tuple));
+    entries.add((BarEntry)getEntryFromTuple(tuple));
   }
 
+  @Override
+  protected boolean areEntriesEqual(Entry e1, Entry e2) {
+    // To avoid (unlikely) cast exceptions, check that
+    // both the entries are of instance BarEntry, and return
+    // false if that is not the case.
+    if (!(e1 instanceof BarEntry && e2 instanceof BarEntry)) {
+      return false;
+    }
+
+    // Due to Bar grouping affecting the x values of the
+    // entries, we need to floor the entries to get rid
+    // of the decimal part. The offset is guaranteed to be
+    // less than 1 (based on the source code and the chosen
+    // granularity and spacing)
+    return Math.floor(e1.getX()) == Math.floor(e2.getX()) &&
+        e1.getY() == e2.getY();
+  }
+
+  @Override
+  public YailList getTupleFromEntry(Entry entry) {
+    // Create a list with the X and Y values of the entry, and
+    // convert the generic List to a YailList. Since Bar Chart
+    // grouping adds offsets to the x value (which are expected
+    // to be below 1), the x value needs to be floored.
+    List tupleEntries = Arrays.asList(Math.floor(entry.getX()), entry.getY());
```

Comment:
Use Math.floor instead of Math.round when getting entry from tuple with decimal x value